File: app/src/main/java/com/lxm/test/utils/ActivityManager.java
package com.lxm.test.utils;

import android.app.Activity;
import java.util.Stack;

/**
 * @Copyright (C), 2020
 * @Author: ym
 * @Date:
 * @Description:
 */
public class ActivityManager {

    private static volatile Stack<Activity> activityStack;
    private static volatile ActivityManager sInstance;

    private ActivityManager() {
    }

    public static synchronized ActivityManager getInstance() {
        if (sInstance == null) {
            synchronized (ActivityManager.class) {
                if (sInstance == null) {
                    sInstance = new ActivityManager();
                    if (activityStack == null) {
                        activityStack = new Stack<>();
                    }
                }
            }
        }
        return sInstance;
    }

    public int size() {
        return activityStack.size();
    }

    public void popActivity() {
        if (size() > 0) {
            Activity activity = activityStack.lastElement();
            if (activity != null) {
                activity.finish();
                activityStack.remove(activity);
            }
        }
    }

    public void popActivity(Activity activity) {
        if (activity != null) {
            activity.finish();
            activityStack.remove(activity);
        }
    }

    public Activity currentActivity() {
        Activity activity = null;
        if (activityStack != null && !activityStack.isEmpty()) {
            activity = activityStack.lastElement();
        }
        return activity;
    }

    public void pushActivity(Activity activity) {
        if (activityStack == null) {
            activityStack = new Stack<>();
        }
        activityStack.add(activity);
    }

    public void popAllActivityExceptOne(Class<?> cls) {
        if (!activityStack.isEmpty()) {
            int stockLen = activityStack.size() - 1;
            for (int i = stockLen; i >= 0; i--) {
                Activity activity = activityStack.get(i);
                if (null != activity) {
                    if (cls != null && activity.getClass().equals(cls)) {
                    } else {
                        popActivity(activity);
                    }
                }
            }
        }
    }

    public void popAllActivity(String... clsStr) {
        int count = 0;
        if (!activityStack.isEmpty()) {
            int stockLen = activityStack.size() - 1;
            for (int i = stockLen; i > 0; i--) {
                Activity activity = activityStack.get(i);
                if (activity == null) {
                    break;
                }
                int len = clsStr.length;
                for (String aClsStr : clsStr) {
                    if (activity.getClass().getSimpleName().equals(aClsStr)) {
                        popActivity(activity);
                        count++;
                    }
                }
                if (count == len) {
                    break;
                }
            }
        }
    }

    public void popAllActivity() {
        popAllActivityExceptOne(null);
    }

    public void popAllActivity(Class<?> cls) {
        popAllActivityExceptOne(cls);
    }
}

File: app/src/main/java/com/lxm/test/utils/ActivityManager.kt
package com.lxm.test.utils


import android.app.Activity
import java.lang.ref.WeakReference
import java.util.*

/**
 * @author : lxm
 * @package_name ：com.jganzhiyun.controlbox.uitls
 * @date : 2021/3/18 16:25
 * @description ：activity管理
 */
object ActivityManager {

    private var activityStack: Stack<Activity>? = null
    var currentActivity: WeakReference<Activity>? = null


    @Synchronized
    fun getCurrentActivity(): Activity? {
        var activity: Activity? = null
        if (currentActivity != null) {
            activity = currentActivity?.get()
        }
        return activity
    }

    @Synchronized
    fun setCurrentActivity(activity: Activity) {
        currentActivity = WeakReference(activity)
    }

    @Synchronized
    fun popActivity(activity: Activity) {
        activity.finish()
        activityStack?.remove(activity)
    }

    @Synchronized
    fun pushActivity(activity: Activity) {
        if (activityStack == null) {
            activityStack = Stack()
        }
        activityStack?.add(activity)
    }

    @Synchronized
    fun takeActivity(clazz: Class<*>) : Activity? {
        activityStack?.let {
            for ((index,value) in it.withIndex()){
                  if(clazz == value::class)
                      return value
            }
        }
        return null
    }

    private fun popAllActivityExceptOne(clazz: Class<Any>?) {
        activityStack?.isNotEmpty()?.let {
            if (it) {
                activityStack?.size?.run {
                    val length = this
                    for (i in length - 1 downTo 0) {
                        activityStack?.run {
                            val activity: Activity = this[i]
                            if (activity.javaClass != clazz)
                                popActivity(activity)
                        }

                    }
                }

            }
        }
    }

    fun popAllActivity(vararg clsStr: String) {
        var count = 0
        activityStack?.isNotEmpty()?.let {
            if (it) {
                activityStack?.size?.run {
                    val stockLen: Int = this - 1
                    for (i in stockLen downTo 1) {
                        val activity: Activity =
                            activityStack!![i] ?: break
                        val len = clsStr.size
                        for (aClsStr in clsStr) {
                            if (activity.javaClass.simpleName == aClsStr) {
                                popActivity(activity)
                                count++
                            }
                        }
                        if (count == len)
                            break
                    }
                }
            }
        }
    }

    fun popAllActivity() {
        popAllActivityExceptOne(null)
    }

    fun getSize(): Int? {
        return activityStack?.size
    }

}